integration/auth/auth_test.go

```diff
@@ -260,6 +265,85 @@ func signInWithPassword(email, password string) (string, error) {
 	return respBody.IDToken, err
 }
 
+func TestSaltPasswordOrder(t *testing.T) {
+	uid := randomUID()
+	email := randomEmail(uid)
+	password := "pass123123"
+	key := "skeleton"
+	salt := "NaCl"
+	tests := []struct {
+		testName   string
+		hashConfig auth.UserImportHash
+		localHash  officialHash.Hash
+	}{
+		{
+			testName: "SHA1_SaltFirst",
+			hashConfig: hash.SHA1{
+				Rounds:     1,
+				InputOrder: hash.InputOrderSaltFirst,
+			},
+			localHash: sha1.New(),
+		},
+		{
+			testName: "HMAC_SHA256_PasswordFirst",
+			hashConfig: hash.HMACSHA256{
+				Key:        []byte(key),
+				InputOrder: hash.InputOrderPasswordFirst,
+			},
+			localHash: hmac.New(sha256.New, []byte(key)),
+		},
+	}
+	// h := hash.HMACSHA256{
+	// 	Key: []byte(key),
+	// 	InputOrder: InputOrderSaltFirst
+	// }
+	// myHash := hmac.New(sha256.New, []byte(key))
+	// myHash.Write([]byte(salt+password))
+	for _, test := range tests {
+		hC, _ := test.hashConfig.Config()
```

Comment:
We should avoid using internal APIs in integration tests. You can include a bool or an InputOrderType value in the test config struct instead.

```
if (test.passwordFirst) {

} else {
  
}
```

Or (possibly better), turn `test.localHash` into a `func() []byte`:

```
{
  localHash: func() []byte {
    h := sha1.New()
    h.Write([]byte(password + salt)
    return h.Sum(nil)
  }
}

```

Reply:
Do you want me to manually specify the results of hashConfig.Config instead, then?  I thought that I should be using our own functions to test that this actually works (or should I only use those in non-integration tests?)

Reply:
Yes, internal APIs are ok for unit tests where we already employ a mocks and other tricks to manipulate the API behavior. These integration tests should be implemented more or less from the developer's standpoint without calling any internal APIs.